// @linaria/rollup@3.0.0-beta.23/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = linaria;

var _path = _interopRequireDefault(require("path"));

var _pluginutils = require("@rollup/pluginutils");

var _babelPreset = require("@linaria/babel-preset");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This file contains a Rollup loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
function linaria({
  include,
  exclude,
  sourceMap,
  preprocessor,
  ...rest
} = {}) {
  const filter = (0, _pluginutils.createFilter)(include, exclude);
  const cssLookup = {};
  let config;
  return {
    name: 'linaria',

    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },

    load(id) {
      return cssLookup[id];
    },

    /* eslint-disable-next-line consistent-return */
    resolveId(importee) {
      if (importee in cssLookup) return importee;
    },

    transform(code, id) {
      var _config, _config2;

      // Do not transform ignored and generated files
      if (!filter(id) || id in cssLookup) return;
      const result = (0, _babelPreset.transform)(code, {
        filename: id,
        preprocessor,
        pluginOptions: rest
      });
      if (!result.cssText) return;
      let {
        cssText
      } = result;
      const slug = (0, _babelPreset.slugify)(cssText);
      const filename = `${id.replace(/\.[jt]sx?$/, '')}_${slug}.css`;

      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`;
      }

      cssLookup[filename] = cssText;

      if (((_config = config) === null || _config === void 0 ? void 0 : _config.command) === 'serve' && (_config2 = config) !== null && _config2 !== void 0 && _config2.root) {
        cssLookup[`/${_path.default.posix.relative(config.root, filename)}`] = cssText;
      }

      result.code += `\nimport ${JSON.stringify(filename)};\n`;
      /* eslint-disable-next-line consistent-return */

      return {
        code: result.code,
        map: result.sourceMap
      };
    }

  };
}
