// @linaria/rollup@3.0.0-beta.23/esm/index.js
/**
 * This file contains a Rollup loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
import path from 'path';
import { createFilter } from '@rollup/pluginutils';
import { transform, slugify } from '@linaria/babel-preset';
export default function linaria({
  include,
  exclude,
  sourceMap,
  preprocessor,
  ...rest
} = {}) {
  const filter = createFilter(include, exclude);
  const cssLookup = {};
  let config;
  return {
    name: 'linaria',

    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },

    load(id) {
      return cssLookup[id];
    },

    /* eslint-disable-next-line consistent-return */
    resolveId(importee) {
      if (importee in cssLookup) return importee;
    },

    transform(code, id) {
      // Do not transform ignored and generated files
      if (!filter(id) || id in cssLookup) return;
      const result = transform(code, {
        filename: id,
        preprocessor,
        pluginOptions: rest
      });
      if (!result.cssText) return;
      let {
        cssText
      } = result;
      const slug = slugify(cssText);
      const filename = `${id.replace(/\.[jt]sx?$/, '')}_${slug}.css`;

      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`;
      }

      cssLookup[filename] = cssText;

      if (config?.command === 'serve' && config?.root) {
        cssLookup[`/${path.posix.relative(config.root, filename)}`] = cssText;
      }

      result.code += `\nimport ${JSON.stringify(filename)};\n`;
      /* eslint-disable-next-line consistent-return */

      return {
        code: result.code,
        map: result.sourceMap
      };
    }

  };
}
